### ISSUE TITLE
Pin all the dependencies

### ISSUE THREAD

**Issue by 0x2b3bfa0** [iterative/mlem#357]
Currently, most of the `mlem` dependencies aren't being pinned to exact versions, and some of them aren't being pinned at all. Because of this, `mlem` will remain broken until https://github.com/iterative/mlem/pull/349 is released; `typer` wasn't pinned, and the last release included expected breaking changes.

> _[Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.](https://semver.org/#spec-item-4)_

https://github.com/iterative/mlem/blob/2e57e183b6bac5626e722637135da141b17b2f60/setup.py#L10

Please consider using a dependency lock file (e.g. [poetry.lock](https://python-poetry.org)) or manually pinning all the dependencies on `setup.py` if that makes more sense. 🤷🏼‍♂️

**Comment by 0x2b3bfa0** [comment]
(Related to [this thread](https://discord.com/channels/485586884165107732/903647230655881226/997068989656219650) on Discord)

**Comment by aguschin** [comment]
Yes, MLEM will break from time to time with no limitations. Also curious why DVC [doesn't limit upper versions for install_requires](https://github.com/iterative/dvc/blob/main/setup.cfg). cc @skshetry

MLEM is a library, so we need to maintain limitations to be as broad as possible. To my knowledge, this means we need to keep lower and upper bounds for each library. Do you know some holistic way to find out lower and upper limits, @0x2b3bfa0? I guess that includes taking a package, installing all versions, and figuring out which ones let MLEM work. Then repeat for all packages. Doing this from time to time, we could get the exact bounds for each package.

**Comment by mike0sv** [comment]
We deliberately dont pin exact version because of what @aguschin said. I set upper bounds for some of them, but this does not guarantee there will be no errors like that. It makes sense to set upper bounds for major versions, but sometimes maintainers break compatibility even in minor releases. I don't know what is the best approach here - set upper bounds, update them with renovate and hope everything will work or just fix errors as they come

**Comment by 0x2b3bfa0** [comment]
If you want to constrains versions as loosely as possible, you may want to follow https://semver.org sanity guidelines: e.g. major version 0 means that a minor version bump can include breaking changes.
